core/server/services/auth/session/index.js

```diff
@@ -0,0 +1,130 @@
+const {URL} = require('url'),
+    common = require('../../../lib/common'),
+    config = require('../../../config'),
+    settingsCache = require('../../settings/cache'),
+    {User} = require('../../../models/user'),
+    {Session} = require('../../../models/session'),
+    session = require('express-session'),
+    SessionStore = require('./store'),
+    {BadRequestError, UnauthorizedError, InternalServerError} = require('../../../lib/common/errors');
+
+const getOrigin = function getOrigin(req) {
+    const origin = req.get('origin');
+    const referrer = req.get('referrer');
+
+    if (!origin && !referrer) {
+        return null;
+    }
+
+    if (origin) {
+        return origin;
+    }
+
+    try {
+        return new URL(req.get('referrer')).origin;
+    } catch (e) {
+        return null;
+    }
+};
+
+const createSession = function createSession(req, res, next) {
+    if (!req.body) {
+        return next(new UnauthorizedError({
+            message: common.i18n.t('errors.middleware.auth.accessDenied')
+        }));
+    }
+    const origin = getOrigin(req);
+    if (!origin) {
+        return next(new BadRequestError({
+            message: common.i18n.t('errors.middleware.auth.unknownOrigin')
+        }));
+    }
+    const {username, password} = req.body;
+    User.check({
+        email: username,
+        password
+    }).then((user) => {
+        req.session.user_id = user.id;
+        req.session.origin = origin;
+        req.session.user_agent = req.get('user-agent');
+        req.session.ip = req.ip;
+        res.sendStatus(201);
+    }).catch(() => {
+        next(new UnauthorizedError({
+            message: common.i18n.t('errors.middleware.auth.accessDenied')
+        }));
+    });
+};
+
+const destroySession = function destroySession(req, res, next) {
+    req.session.destroy((err) => {
+        if (err) {
+            return next(new InternalServerError());
+        }
+        return res.sendStatus(204);
+    });
+};
+
+const getUser = function getUser(req, res, next) {
+    if (!req.session || !req.session.user_id) {
+        req.user = null;
+        return next();
+    }
+    User.findOne({id: req.session.user_id})
+        .then((user) => {
+            req.user = user;
+            next();
+        }).catch(() => {
+            next(new UnauthorizedError({
+                message: common.i18n.t('errors.middleware.auth.accessDenied')
+            }));
+        });
+};
+
+const ensureUser = function ensureUser(req, res, next) {
+    if (req.user && req.user.id) {
+        return next();
+    }
+    next(new UnauthorizedError({
+        message: common.i18n.t('errors.middleware.auth.accessDenied')
+    }));
+};
+
+const getSession = session({
```

Comment:
The session get's initialized on bootstrap 🤔Is that intended and needed?

Reply:
Yeah it's intended - is there a reason not to?

Reply:
And why is it intended? 🙂

General reasons why we should avoid this:

- we try to avoid instantiating things which doesn't have to on bootstrap (bootstrap === starting the server), because it can slow down bootstrapping
- we prefer initialise on needs/runtime
- we also prefer to load 3rd party libs async on runtime e.g. `require('express-session')` when you need it, because if you start Ghost, any third lib get's required and this get's out of control and some modules really load slow
- it's harder to test this file, because it auto initialises a session (not functional) as soon as this file get's required
- if i require this file, i don't expect an auto initialisation

Happy to hear why you think it is intended. If there is a really good reason doing that, i am fine with leaving it and adding a comment 👍

Reply:
> i'll add an init method and put it behind that if you like?

It's not "just me liking it", i think it's makes more sense because of the mentioned reasons :)

Reply:
> i think it's makes more sense because of the mentioned reasons

I agree, I didn't mean it to come off like that - sorry